File: client/src/physics.ts
/**
 * Client-side physics engine for local play.
 * Mirrors the Rust Rapier2D-based server physics closely enough for single-player/local multiplayer.
 * When online, this will be replaced by server-authoritative state + client prediction via WASM.
 */

import type { TankInput } from "./input";
import type { GameMap } from "./map";

// --- Constants (must match shared/src/physics.rs) ---
export const TANK_WIDTH = 30;
export const TANK_HEIGHT = 20;
export const TANK_SPEED = 150;
export const TANK_ROTATION_SPEED = 3.5;
export const BULLET_RADIUS = 4;
export const BULLET_SPEED = 350;
export const MAX_BULLETS_PER_TANK = 5;
export const BULLET_MAX_BOUNCES = 50;
export const PHYSICS_TIMESTEP = 1 / 60;

// --- Types ---

export interface Tank {
	id: number;
	x: number;
	y: number;
	rotation: number;
	vx: number;
	vy: number;
	alive: boolean;
	bulletCount: number;
	color: string;
}

export interface Bullet {
	id: number;
	x: number;
	y: number;
	vx: number;
	vy: number;
	owner: number;
	bounces: number;
	alive: boolean;
}

export interface GameEvent {
	type: "tank_destroyed" | "bullet_fired" | "bullet_bounced" | "bullet_destroyed";
	data: Record<string, unknown>;
}

// --- Line segment representation for walls ---

interface Segment {
	x1: number;
	y1: number;
	x2: number;
	y2: number;
	nx: number;
	ny: number; // outward normal
}

// --- Physics World ---

const TANK_COLORS = ["#e94560", "#0f3460", "#16c79a", "#f5a623"];

export class PhysicsWorld {
	tanks: Tank[] = [];
	bullets: Bullet[] = [];
	private segments: Segment[] = [];
	private nextBulletId = 0;
	map: GameMap;
	tick = 0;

	constructor(map: GameMap) {
		this.map = map;
		this.buildWallSegments();
	}

	/** Convert thick walls into line segments for collision. */
	private buildWallSegments() {
		for (const wall of this.map.walls) {
			const dx = wall.x2 - wall.x1;
			const dy = wall.y2 - wall.y1;
			const len = Math.sqrt(dx * dx + dy * dy);
			if (len < 0.01) continue;

			// Normal perpendicular to wall direction
			const nx = -dy / len;
			const ny = dx / len;
			const ht = wall.thickness / 2;

			// Two sides of the thick wall
			// Side 1 (offset by +normal * half_thickness)
			this.segments.push({
				x1: wall.x1 + nx * ht,
				y1: wall.y1 + ny * ht,
				x2: wall.x2 + nx * ht,
				y2: wall.y2 + ny * ht,
				nx,
				ny,
			});
			// Side 2 (offset by -normal * half_thickness)
			this.segments.push({
				x1: wall.x1 - nx * ht,
				y1: wall.y1 - ny * ht,
				x2: wall.x2 - nx * ht,
				y2: wall.y2 - ny * ht,
				nx: -nx,
				ny: -ny,
			});

			// End caps (for wall endpoints)
			// Cap at (x1, y1)
			this.segments.push({
				x1: wall.x1 + nx * ht,
				y1: wall.y1 + ny * ht,
				x2: wall.x1 - nx * ht,
				y2: wall.y1 - ny * ht,
				nx: -dx / len,
				ny: -dy / len,
			});
			// Cap at (x2, y2)
			this.segments.push({
				x1: wall.x2 + nx * ht,
				y1: wall.y2 + ny * ht,
				x2: wall.x2 - nx * ht,
				y2: wall.y2 - ny * ht,
				nx: dx / len,
				ny: dy / len,
			});
		}
	}

	addTank(id: number) {
		const spawnIdx = id % this.map.spawnPoints.length;
		const [sx, sy] = this.map.spawnPoints[spawnIdx];
		this.tanks.push({
			id,
			x: sx,
			y: sy,
			rotation: 0,
			vx: 0,
			vy: 0,
			alive: true,
			bulletCount: 0,
			color: TANK_COLORS[id % TANK_COLORS.length],
		});
	}

	step(inputs: Map<number, TankInput>, dt: number): GameEvent[] {
		const events: GameEvent[] = [];

		// --- Update tanks ---
		for (const tank of this.tanks) {
			if (!tank.alive) continue;
			const input = inputs.get(tank.id);
			if (!input) continue;

			// Rotation
			if (input.turnLeft) tank.rotation -= TANK_ROTATION_SPEED * dt;
			if (input.turnRight) tank.rotation += TANK_ROTATION_SPEED * dt;

			// Movement
			let speed = 0;
			if (input.forward) speed += TANK_SPEED;
			if (input.backward) speed -= TANK_SPEED * 0.6;

			const newVx = speed * Math.cos(tank.rotation);
			const newVy = speed * Math.sin(tank.rotation);

			const newX = tank.x + newVx * dt;
			const newY = tank.y + newVy * dt;

			// Collision with walls (simple push-out)
			const resolved = this.resolveTankWallCollision(newX, newY, TANK_WIDTH / 2);
			tank.x = resolved.x;
			tank.y = resolved.y;
			tank.vx = newVx;
			tank.vy = newVy;

			// Collision with other tanks
			for (const other of this.tanks) {
				if (other.id === tank.id || !other.alive) continue;
				const ddx = tank.x - other.x;
				const ddy = tank.y - other.y;
				const dist = Math.sqrt(ddx * ddx + ddy * ddy);
				const minDist = TANK_WIDTH; // simplified: both tanks same radius
				if (dist < minDist && dist > 0.01) {
					const overlap = minDist - dist;
					const pushX = (ddx / dist) * overlap * 0.5;
					const pushY = (ddy / dist) * overlap * 0.5;
					tank.x += pushX;
					tank.y += pushY;
					other.x -= pushX;
					other.y -= pushY;
				}
			}

			// Fire
			if (input.fire && tank.bulletCount < MAX_BULLETS_PER_TANK) {
				const spawnDist = TANK_WIDTH / 2 + BULLET_RADIUS + 2;
				const bx = tank.x + Math.cos(tank.rotation) * spawnDist;
				const by = tank.y + Math.sin(tank.rotation) * spawnDist;
				const bvx = BULLET_SPEED * Math.cos(tank.rotation);
				const bvy = BULLET_SPEED * Math.sin(tank.rotation);
				const bulletId = this.nextBulletId++;

				this.bullets.push({
					id: bulletId,
					x: bx,
					y: by,
					vx: bvx,
					vy: bvy,
					owner: tank.id,
					bounces: 0,
					alive: true,
				});
				tank.bulletCount++;
				events.push({ type: "bullet_fired", data: { owner: tank.id, x: bx, y: by } });
			}
		}

		// --- Update bullets ---
		for (const bullet of this.bullets) {
			if (!bullet.alive) continue;

			const newX = bullet.x + bullet.vx * dt;
			const newY = bullet.y + bullet.vy * dt;

			// Check wall collisions (reflect)
			const wallHit = this.bulletWallCollision(bullet.x, bullet.y, newX, newY, BULLET_RADIUS);
			if (wallHit) {
				bullet.x = wallHit.x;
				bullet.y = wallHit.y;
				bullet.vx = wallHit.vx;
				bullet.vy = wallHit.vy;
				bullet.bounces++;
				events.push({ type: "bullet_bounced", data: { bulletId: bullet.id } });

				if (bullet.bounces >= BULLET_MAX_BOUNCES) {
					bullet.alive = false;
					events.push({ type: "bullet_destroyed", data: { bulletId: bullet.id } });
				}
			} else {
				bullet.x = newX;
				bullet.y = newY;
			}

			// Check tank collisions
			for (const tank of this.tanks) {
				if (!tank.alive) continue;
				const ddx = bullet.x - tank.x;
				const ddy = bullet.y - tank.y;
				const dist = Math.sqrt(ddx * ddx + ddy * ddy);
				if (dist < TANK_WIDTH / 2 + BULLET_RADIUS) {
					tank.alive = false;
					bullet.alive = false;
					events.push({ type: "tank_destroyed", data: { victim: tank.id, killer: bullet.owner } });
					events.push({ type: "bullet_destroyed", data: { bulletId: bullet.id } });
				}
			}
		}

		// Clean up dead bullets
		for (const bullet of this.bullets) {
			if (!bullet.alive) {
				const owner = this.tanks.find((t) => t.id === bullet.owner);
				if (owner && owner.bulletCount > 0) owner.bulletCount--;
			}
		}
		this.bullets = this.bullets.filter((b) => b.alive);

		this.tick++;
		return events;
	}

	/** Simple wall collision for circular tank. Push out of walls. */
	private resolveTankWallCollision(x: number, y: number, radius: number): { x: number; y: number } {
		for (const wall of this.map.walls) {
			const result = pointToSegmentDistance(x, y, wall.x1, wall.y1, wall.x2, wall.y2);
			const minDist = radius + wall.thickness / 2;
			if (result.dist < minDist) {
				const push = minDist - result.dist;
				if (result.dist > 0.01) {
					x += result.nx * push;
					y += result.ny * push;
				}
			}
		}
		return { x, y };
	}

	/** Bullet-wall collision with reflection. */
	private bulletWallCollision(
		x0: number,
		y0: number,
		x1: number,
		y1: number,
		radius: number,
	): { x: number; y: number; vx: number; vy: number } | null {
		let closestT = Infinity;
		let hitNormal = { x: 0, y: 0 };

		for (const wall of this.map.walls) {
			// Check distance from new position to wall segment
			const result = pointToSegmentDistance(x1, y1, wall.x1, wall.y1, wall.x2, wall.y2);
			const minDist = radius + wall.thickness / 2;
			if (result.dist < minDist) {
				// Find approximate collision time
				const result0 = pointToSegmentDistance(x0, y0, wall.x1, wall.y1, wall.x2, wall.y2);
				if (result0.dist >= minDist) {
					const t = (result0.dist - minDist) / (result0.dist - result.dist);
					if (t < closestT) {
						closestT = t;
						hitNormal = { x: result.nx, y: result.ny };
					}
				} else if (result.dist < closestT) {
					closestT = 0;
					hitNormal = { x: result.nx, y: result.ny };
				}
			}
		}

		if (closestT <= 1) {
			const dx = x1 - x0;
			const dy = y1 - y0;
			// Move to just before collision
			const cx = x0 + dx * Math.min(closestT, 1);
			const cy = y0 + dy * Math.min(closestT, 1);
			// Reflect velocity
			const dot = dx * hitNormal.x + dy * hitNormal.y;
			const rvx = dx - 2 * dot * hitNormal.x;
			const rvy = dy - 2 * dot * hitNormal.y;
			// Normalize to bullet speed
			const speed = Math.sqrt(rvx * rvx + rvy * rvy);
			return {
				x: cx + hitNormal.x * 1, // push out slightly
				y: cy + hitNormal.y * 1,
				vx: (rvx / PHYSICS_TIMESTEP) * (speed > 0.01 ? BULLET_SPEED / speed : 1),
				vy: (rvy / PHYSICS_TIMESTEP) * (speed > 0.01 ? BULLET_SPEED / speed : 1),
			};
		}
		return null;
	}

	/** Check if only one (or zero) tanks are alive. */
	checkRoundOver(): number | null {
		if (this.tanks.length <= 1) return null;
		const alive = this.tanks.filter((t) => t.alive);
		if (alive.length <= 1) {
			return alive.length === 1 ? alive[0].id : -1; // -1 = draw
		}
		return null;
	}
}

/** Distance from point to line segment, with outward normal. */
function pointToSegmentDistance(
	px: number,
	py: number,
	x1: number,
	y1: number,
	x2: number,
	y2: number,
): { dist: number; nx: number; ny: number } {
	const dx = x2 - x1;
	const dy = y2 - y1;
	const lenSq = dx * dx + dy * dy;

	let t = 0;
	if (lenSq > 0.0001) {
		t = Math.max(0, Math.min(1, ((px - x1) * dx + (py - y1) * dy) / lenSq));
	}

	const closestX = x1 + t * dx;
	const closestY = y1 + t * dy;
	const diffX = px - closestX;
	const diffY = py - closestY;
	const dist = Math.sqrt(diffX * diffX + diffY * diffY);

	if (dist < 0.0001) {
		// Point is on the segment — use perpendicular normal
		const len = Math.sqrt(lenSq);
		return { dist: 0, nx: -dy / len, ny: dx / len };
	}

	return { dist, nx: diffX / dist, ny: diffY / dist };
}

File: shared/src/physics.rs
use rapier2d::prelude::*;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

use crate::map::GameMap;

// --- Constants ---

pub const TANK_WIDTH: f32 = 30.0;
pub const TANK_HEIGHT: f32 = 20.0;
pub const TANK_SPEED: f32 = 150.0;
pub const TANK_ROTATION_SPEED: f32 = 3.5;
pub const BULLET_RADIUS: f32 = 4.0;
pub const BULLET_SPEED: f32 = 350.0;
pub const MAX_BULLETS_PER_TANK: usize = 5;
pub const BULLET_MAX_BOUNCES: u32 = 50;
pub const PHYSICS_TIMESTEP: f32 = 1.0 / 60.0;

// --- Input ---

#[derive(Debug, Clone, Copy, Default, Serialize, Deserialize)]
pub struct TankInput {
    pub forward: bool,
    pub backward: bool,
    pub turn_left: bool,
    pub turn_right: bool,
    pub fire: bool,
}

// --- Identifiers ---

pub type PlayerId = u32;

// --- Collision groups ---

const WALL_GROUP: u32 = 0b0001;
const TANK_GROUP: u32 = 0b0010;
const BULLET_GROUP: u32 = 0b0100;

fn wall_collision_groups() -> InteractionGroups {
    InteractionGroups::new(Group::from_bits_truncate(WALL_GROUP), Group::all())
}

fn tank_collision_groups() -> InteractionGroups {
    InteractionGroups::new(
        Group::from_bits_truncate(TANK_GROUP),
        Group::from_bits_truncate(WALL_GROUP | TANK_GROUP),
    )
}

fn bullet_collision_groups() -> InteractionGroups {
    InteractionGroups::new(
        Group::from_bits_truncate(BULLET_GROUP),
        Group::from_bits_truncate(WALL_GROUP | TANK_GROUP),
    )
}

// --- Tank state ---

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TankState {
    pub id: PlayerId,
    pub x: f32,
    pub y: f32,
    pub rotation: f32,
    pub alive: bool,
    pub bullet_count: usize,
}

// --- Bullet state ---

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BulletState {
    pub id: u32,
    pub x: f32,
    pub y: f32,
    pub vx: f32,
    pub vy: f32,
    pub owner: PlayerId,
    pub bounces: u32,
}

// --- Internal tracking ---

struct TankBody {
    body_handle: RigidBodyHandle,
    collider_handle: ColliderHandle,
    alive: bool,
    bullet_count: usize,
}

struct BulletBody {
    id: u32,
    body_handle: RigidBodyHandle,
    collider_handle: ColliderHandle,
    owner: PlayerId,
    bounces: u32,
    remove: bool,
}

// --- Game events ---

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum GameEvent {
    TankDestroyed { victim: PlayerId, killer: PlayerId },
    BulletFired { owner: PlayerId, x: f32, y: f32, vx: f32, vy: f32 },
    BulletBounced { bullet_id: u32 },
    BulletDestroyed { bullet_id: u32 },
}

// --- Snapshot for network sync ---

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GameSnapshot {
    pub tick: u64,
    pub tanks: Vec<TankState>,
    pub bullets: Vec<BulletState>,
}

// --- Physics world ---

pub struct PhysicsWorld {
    rigid_body_set: RigidBodySet,
    collider_set: ColliderSet,
    integration_parameters: IntegrationParameters,
    physics_pipeline: PhysicsPipeline,
    island_manager: IslandManager,
    broad_phase: DefaultBroadPhase,
    narrow_phase: NarrowPhase,
    impulse_joint_set: ImpulseJointSet,
    multibody_joint_set: MultibodyJointSet,
    ccd_solver: CCDSolver,

    tanks: HashMap<PlayerId, TankBody>,
    bullets: Vec<BulletBody>,
    _wall_colliders: Vec<ColliderHandle>,
    next_bullet_id: u32,
    pub tick: u64,

    map: GameMap,
}

impl PhysicsWorld {
    pub fn new(map: GameMap) -> Self {
        let mut rigid_body_set = RigidBodySet::new();
        let mut collider_set = ColliderSet::new();

        let integration_parameters = IntegrationParameters {
            dt: PHYSICS_TIMESTEP,
            ..Default::default()
        };

        let mut wall_colliders = Vec::new();

        // Create wall colliders
        for wall in &map.walls {
            let cx = (wall.x1 + wall.x2) / 2.0;
            let cy = (wall.y1 + wall.y2) / 2.0;
            let dx = wall.x2 - wall.x1;
            let dy = wall.y2 - wall.y1;
            let length = (dx * dx + dy * dy).sqrt();
            let angle = dy.atan2(dx);

            let body = RigidBodyBuilder::fixed()
                .translation(vector![cx, cy])
                .rotation(angle)
                .build();
            let body_handle = rigid_body_set.insert(body);

            let half_thickness = wall.thickness / 2.0;
            let half_length = length / 2.0;

            let collider = ColliderBuilder::cuboid(half_length, half_thickness)
                .restitution(1.0) // Perfect bounce for bullets
                .friction(0.0)
                .collision_groups(wall_collision_groups())
                .build();
            let collider_handle = collider_set.insert_with_parent(collider, body_handle, &mut rigid_body_set);
            wall_colliders.push(collider_handle);
        }

        PhysicsWorld {
            rigid_body_set,
            collider_set,
            integration_parameters,
            physics_pipeline: PhysicsPipeline::new(),
            island_manager: IslandManager::new(),
            broad_phase: DefaultBroadPhase::new(),
            narrow_phase: NarrowPhase::new(),
            impulse_joint_set: ImpulseJointSet::new(),
            multibody_joint_set: MultibodyJointSet::new(),
            ccd_solver: CCDSolver::new(),
            tanks: HashMap::new(),
            bullets: Vec::new(),
            _wall_colliders: wall_colliders,
            next_bullet_id: 0,
            tick: 0,
            map,
        }
    }

    pub fn add_tank(&mut self, id: PlayerId) {
        let spawn_idx = id as usize % self.map.spawn_points.len();
        let (sx, sy) = self.map.spawn_points[spawn_idx];

        let body = RigidBodyBuilder::dynamic()
            .translation(vector![sx, sy])
            .linear_damping(10.0) // High damping so tank stops quickly
            .angular_damping(10.0)
            .build();
        let body_handle = self.rigid_body_set.insert(body);

        let collider = ColliderBuilder::cuboid(TANK_WIDTH / 2.0, TANK_HEIGHT / 2.0)
            .density(1.0)
            .friction(0.0)
            .restitution(0.0)
            .collision_groups(tank_collision_groups())
            .active_events(ActiveEvents::COLLISION_EVENTS)
            .build();
        let collider_handle = self.collider_set.insert_with_parent(
            collider,
            body_handle,
            &mut self.rigid_body_set,
        );

        self.tanks.insert(id, TankBody {
            body_handle,
            collider_handle,
            alive: true,
            bullet_count: 0,
        });
    }

    pub fn remove_tank(&mut self, id: PlayerId) {
        if let Some(tank) = self.tanks.remove(&id) {
            self.rigid_body_set.remove(
                tank.body_handle,
                &mut self.island_manager,
                &mut self.collider_set,
                &mut self.impulse_joint_set,
                &mut self.multibody_joint_set,
                true,
            );
        }
    }

    /// Apply input and step the physics world. Returns events from this tick.
    pub fn step(&mut self, inputs: &HashMap<PlayerId, TankInput>) -> Vec<GameEvent> {
        let mut events = Vec::new();

        // Apply tank inputs
        for (&id, input) in inputs {
            if let Some(tank) = self.tanks.get(&id) {
                if !tank.alive {
                    continue;
                }
                if let Some(body) = self.rigid_body_set.get_mut(tank.body_handle) {
                    let rotation = body.rotation().angle();

                    // Movement
                    let mut force = 0.0;
                    if input.forward { force += TANK_SPEED; }
                    if input.backward { force -= TANK_SPEED * 0.6; }

                    let fx = force * rotation.cos();
                    let fy = force * rotation.sin();
                    body.set_linvel(vector![fx, fy], true);

                    // Rotation
                    let mut torque = 0.0;
                    if input.turn_left { torque -= TANK_ROTATION_SPEED; }
                    if input.turn_right { torque += TANK_ROTATION_SPEED; }
                    body.set_angvel(torque, true);
                }
            }
        }

        // Handle firing
        for (&id, input) in inputs {
            if input.fire
                && let Some(tank) = self.tanks.get(&id) {
                    if !tank.alive || tank.bullet_count >= MAX_BULLETS_PER_TANK {
                        continue;
                    }
                    if let Some(body) = self.rigid_body_set.get(tank.body_handle) {
                        let pos = *body.translation();
                        let angle = body.rotation().angle();

                        // Spawn bullet at tank barrel tip
                        let spawn_dist = TANK_WIDTH / 2.0 + BULLET_RADIUS + 2.0;
                        let bx = pos.x + angle.cos() * spawn_dist;
                        let by = pos.y + angle.sin() * spawn_dist;
                        let vx = BULLET_SPEED * angle.cos();
                        let vy = BULLET_SPEED * angle.sin();

                        let bullet_id = self.next_bullet_id;
                        self.next_bullet_id += 1;

                        let bullet_body = RigidBodyBuilder::dynamic()
                            .translation(vector![bx, by])
                            .linvel(vector![vx, vy])
                            .gravity_scale(0.0)
                            .ccd_enabled(true) // Continuous collision detection — bullet won't tunnel
                            .build();
                        let bh = self.rigid_body_set.insert(bullet_body);

                        let bullet_collider = ColliderBuilder::ball(BULLET_RADIUS)
                            .restitution(1.0)
                            .restitution_combine_rule(CoefficientCombineRule::Max)
                            .friction(0.0)
                            .density(0.1)
                            .collision_groups(bullet_collision_groups())
                            .active_events(ActiveEvents::COLLISION_EVENTS | ActiveEvents::CONTACT_FORCE_EVENTS)
                            .build();
                        let ch = self.collider_set.insert_with_parent(bullet_collider, bh, &mut self.rigid_body_set);

                        self.bullets.push(BulletBody {
                            id: bullet_id,
                            body_handle: bh,
                            collider_handle: ch,
                            owner: id,
                            bounces: 0,
                            remove: false,
                        });

                        if let Some(tank) = self.tanks.get_mut(&id) {
                            tank.bullet_count += 1;
                        }

                        events.push(GameEvent::BulletFired { owner: id, x: bx, y: by, vx, vy });
                    }
                }
        }

        // Step physics
        self.physics_pipeline.step(
            &vector![0.0, 0.0], // No gravity for top-down game
            &self.integration_parameters,
            &mut self.island_manager,
            &mut self.broad_phase,
            &mut self.narrow_phase,
            &mut self.rigid_body_set,
            &mut self.collider_set,
            &mut self.impulse_joint_set,
            &mut self.multibody_joint_set,
            &mut self.ccd_solver,
            None,
            &(),
            &(),
        );

        // Process bullet-wall bounces: re-normalize bullet speed after bounce
        for bullet in &mut self.bullets {
            if bullet.remove {
                continue;
            }
            if let Some(body) = self.rigid_body_set.get_mut(bullet.body_handle) {
                let vel = *body.linvel();
                let speed = (vel.x * vel.x + vel.y * vel.y).sqrt();

                // If speed changed significantly, a bounce happened
                if (speed - BULLET_SPEED).abs() > 1.0 {
                    // Re-normalize to constant speed
                    if speed > 0.01 {
                        let scale = BULLET_SPEED / speed;
                        body.set_linvel(vector![vel.x * scale, vel.y * scale], true);
                    }
                    bullet.bounces += 1;
                    events.push(GameEvent::BulletBounced { bullet_id: bullet.id });

                    if bullet.bounces >= BULLET_MAX_BOUNCES {
                        bullet.remove = true;
                        events.push(GameEvent::BulletDestroyed { bullet_id: bullet.id });
                    }
                }
            }
        }

        // Check bullet-tank collisions
        for pair in self.narrow_phase.contact_pairs() {
            if !pair.has_any_active_contact {
                continue;
            }

            let c1 = pair.collider1;
            let c2 = pair.collider2;

            // Find if one is a bullet and the other is a tank
            let bullet_idx = self.bullets.iter().position(|b| b.collider_handle == c1 || b.collider_handle == c2);
            if let Some(bi) = bullet_idx {
                let bullet_ch = self.bullets[bi].collider_handle;
                let other_ch = if bullet_ch == c1 { c2 } else { c1 };

                // Check if other is a tank
                for (&tank_id, tank) in &mut self.tanks {
                    if tank.collider_handle == other_ch && tank.alive {
                        tank.alive = false;
                        let killer = self.bullets[bi].owner;
                        self.bullets[bi].remove = true;
                        events.push(GameEvent::TankDestroyed { victim: tank_id, killer });
                        events.push(GameEvent::BulletDestroyed { bullet_id: self.bullets[bi].id });
                        break;
                    }
                }
            }
        }

        // Remove dead bullets
        let mut removed_owners = Vec::new();
        self.bullets.retain(|bullet| {
            if bullet.remove {
                removed_owners.push(bullet.owner);
                // Remove from physics
                // We can't borrow self mutably here, so collect and remove after
                false
            } else {
                true
            }
        });

        // Actually remove bullet bodies (we need a second pass since retain doesn't give mutable self)
        // The bodies are already detached from the bullets vec; clean them up
        // Note: we need to track body handles before retain. Let's refactor:
        // Actually the retain already dropped them from our tracking. The rigid bodies remain in the set.
        // We'll clean up orphaned bodies:
        // For simplicity, let's handle this properly:

        // Decrement bullet counts for removed bullets
        for owner in &removed_owners {
            if let Some(tank) = self.tanks.get_mut(owner)
                && tank.bullet_count > 0 {
                    tank.bullet_count -= 1;
                }
        }

        self.tick += 1;
        events
    }

    /// Remove physics bodies for bullets marked for removal.
    /// Call after step() to clean up.
    pub fn cleanup_removed_bullets(&mut self) {
        // Collect handles of bullets to remove
        let to_remove: Vec<RigidBodyHandle> = self.bullets
            .iter()
            .filter(|b| b.remove)
            .map(|b| b.body_handle)
            .collect();

        for handle in to_remove {
            self.rigid_body_set.remove(
                handle,
                &mut self.island_manager,
                &mut self.collider_set,
                &mut self.impulse_joint_set,
                &mut self.multibody_joint_set,
                true,
            );
        }

        self.bullets.retain(|b| !b.remove);
    }

    pub fn snapshot(&self) -> GameSnapshot {
        let tanks = self.tanks.iter().map(|(&id, tank)| {
            let body = &self.rigid_body_set[tank.body_handle];
            let pos = body.translation();
            TankState {
                id,
                x: pos.x,
                y: pos.y,
                rotation: body.rotation().angle(),
                alive: tank.alive,
                bullet_count: tank.bullet_count,
            }
        }).collect();

        let bullets = self.bullets.iter().map(|bullet| {
            let body = &self.rigid_body_set[bullet.body_handle];
            let pos = body.translation();
            let vel = body.linvel();
            BulletState {
                id: bullet.id,
                x: pos.x,
                y: pos.y,
                vx: vel.x,
                vy: vel.y,
                owner: bullet.owner,
                bounces: bullet.bounces,
            }
        }).collect();

        GameSnapshot { tick: self.tick, tanks, bullets }
    }

    pub fn map(&self) -> &GameMap {
        &self.map
    }

    /// Check if all tanks except one (or zero) are dead.
    pub fn check_round_over(&self) -> Option<PlayerId> {
        let alive: Vec<PlayerId> = self.tanks.iter()
            .filter(|(_, t)| t.alive)
            .map(|(&id, _)| id)
            .collect();
        if alive.len() <= 1 && self.tanks.len() > 1 {
            alive.into_iter().next()
        } else {
            None
        }
    }
}
